netket/models/autoreg.py

```diff
@@ -95,10 +104,73 @@ def conditionals(self, inputs: Array) -> Array:
           # it takes probability 0.3 to be spin down (local state index 0),
           # and probability 0.7 to be spin up (local state index 1).
         """
+        if inputs.ndim == 1:
+            inputs = jnp.expand_dims(inputs, axis=0)
+
+        log_psi = self.conditionals_log_psi(inputs)
+
+        p = jnp.exp(self.machine_pow * log_psi.real)
+        return p
+
+    def conditional(self, inputs: Array, index: int) -> Array:
+        """
+        Computes the conditional probabilities for a site to take a given value.
+
+        It should only be called successively with indices 0, 1, 2, ...,
+        as in the autoregressive sampling procedure.
+
+        Args:
+          inputs: configurations with dimensions (batch, Hilbert.size).
+          index: index of the site.
+
+        Returns:
+          The probabilities with dimensions (batch, Hilbert.local_size).
+        """
+        return self.conditionals(inputs)[:, index, :]
+
+    def __call__(self, inputs: Array) -> Array:
+        """Returns log_psi."""
+
+        if inputs.ndim == 1:
+            inputs = jnp.expand_dims(inputs, axis=0)
+
+        idx = self.hilbert.states_to_local_indices(inputs)
+        idx = jnp.expand_dims(idx, axis=-1)
+
+        log_psi = self.conditionals_log_psi(inputs)
+
+        log_psi = jnp.take_along_axis(log_psi, idx, axis=-1)
+        log_psi = log_psi.reshape((inputs.shape[0], -1)).sum(axis=1)
+        return log_psi
+
+
+class AbstractRecursiveARNN(AbstractARNN):
+    def conditionals_log_psi(self, inputs: Array) -> Array:
+        """
+        Implementation of a ARNN that repeatedly calls a set of layers.
+        """
+        inputs = self._reshape_inputs(inputs)
+
+        x = jnp.expand_dims(inputs, axis=-1)
+
+        for i in range(self.layers):
+            if i > 0:
+                x = self.activation(x)
+            x = self._layers[i](x)
+
+        x = x.reshape((x.shape[0], -1, x.shape[-1]))
+        log_psi = _normalize(x, self.machine_pow)
+        return log_psi
+
+    def _reshape_inputs(model: Any, inputs: Array) -> Array:  # noqa: F811
+        """
+        ???
```

Comment:
Suggested docstring: """Reshape the inputs from (batch_size, hilbert_size) to (batch_size, spatial_dims...) before sending them to the ARNN layers."""